src/UNIVesting.sol

```diff
@@ -0,0 +1,119 @@
+  }
+
+  /// @inheritdoc IUNIVesting
+  /// @dev Can only be called by the owner and only when no active quarters are available to
+  /// withdraw (i.e., quarters() == 0). This prevents changing the amount when tokens have already
+  /// vested and are waiting to be claimed
+  function updateVestingAmount(uint256 amount) public onlyOwner {
+    require(quarters() == 0, CannotUpdateAmount());
+    quarterlyVestingAmount = amount;
+    emit VestingAmountUpdated(amount);
+  }
+
+  /// @inheritdoc IUNIVesting
+  /// @dev Both the owner and current recipient can update the recipient address.
+  /// This allows the recipient to transfer their vesting rights to another address.
+  function updateRecipient(address _recipient) public onlyOwnerOrRecipient {
+    recipient = _recipient;
+    emit RecipientUpdated(recipient);
+  }
+
+  /// @inheritdoc IUNIVesting
+  /// @dev This function can be called by anyone (not just the recipient).
+  /// Handles both full and partial withdrawals based on the owner's allowance.
+  /// If allowance < vested amount, only withdraws what's allowed and updates
+  /// the timestamp accordingly. The remaining quarters can be withdrawn later.
+  /// Relies on owner maintaining sufficient ERC20 allowance.
+  /// If owner revokes allowance, withdrawals will fail with InsufficientAllowance.
+  function withdraw() external {
+    uint48 quartersPassed = quarters();
+    require(quartersPassed > 0, OnlyQuarterly());
+
+    uint256 vestedAmount = quarterlyVestingAmount * uint256(quartersPassed);
+    uint256 currentAllowance = UNI.allowance(owner, address(this));
+
+    if (currentAllowance < vestedAmount) {
+      // Partial withdrawal path: owner's allowance is less than vested amount
+      // Calculate how many complete quarters we can withdraw with current allowance
+      uint48 withdrawableQuarters = uint48(currentAllowance / quarterlyVestingAmount);
+
+      require(withdrawableQuarters > 0, InsufficientAllowance());
+
+      // Only advance timestamp by the quarters we're actually paying out
+      // This ensures remaining quarters can be withdrawn later when allowance increases
+      lastUnlockTimestamp =
+        uint48(DateTime.addMonths(lastUnlockTimestamp, withdrawableQuarters * MONTHS_PER_QUARTER));
+
+      vestedAmount = quarterlyVestingAmount * uint256(withdrawableQuarters);
+    } else {
+      // Full withdrawal path: sufficient allowance for all vested quarters
+      // Advance timestamp by all quarters that have vested
+      lastUnlockTimestamp =
+        uint48(DateTime.addMonths(lastUnlockTimestamp, quartersPassed * MONTHS_PER_QUARTER));
+    }
+    UNI.safeTransferFrom(owner, recipient, vestedAmount);
+  }
+
+  /// @inheritdoc IUNIVesting
+  /// @dev Uses calendar-based quarters (3 months each)
+  /// Returns 0 if no quarters have passed since last withdrawal.
+  function quarters() public view returns (uint48 quartersPassed) {
```

Comment:
nit: this function could also be named better

Reply:
or give me a suggestion?